presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -134,94 +194,377 @@ public boolean needsInput()
+             * Let LookupSourceFactory know LookupSources can be disposed as far as we're concerned.
+             */
+            finishRegularInput();
+            unspilling = true;
+        }
+
+        if (probe == null && unspilling && !finished) {
+            /*
+             * If no current partition or it was exhausted, unspill next one.
+             * Add input there when it needs one, produce output. Be Happy.
+             */
+            tryUnspillNext();
+        }
+
+        if (probe != null) {
+            processProbe();
+        }
+
+        return producePage();
+    }
+
+    private void finishRegularInput()
+    {
+        verify(partitionedConsumption == null, "partitioned consumption already started");
+        partitionedConsumption = lookupSourceFactory.finishProbeOperator(lookupJoinsCount);
+    }
+
+    private void tryUnspillNext()
+    {
+        verify(probe == null);
+
+        if (!partitionedConsumption.isDone()) {
+            return;
+        }
+
+        if (lookupPartitions == null) {
+            lookupPartitions = getDone(partitionedConsumption).beginConsumption();
+        }
+
+        if (unspilledInputPages.hasNext()) {
+            addInput(unspilledInputPages.next());
+            verify(probe != null);
+            return;
+        }
+
+        if (unspilledLookupSource.isPresent()) {
+            if (!unspilledLookupSource.get().isDone()) {
+                // Not unspilled yet
+                return;
+            }
+            LookupSource lookupSource = getDone(unspilledLookupSource.get()).get();
+            unspilledLookupSource = Optional.empty();
+
+            // Close previous lookupSourceProvider (either supplied initially or for the previous partition)
+            lookupSourceProvider.close();
+            lookupSourceProvider = new StaticLookupSourceProvider(lookupSource);
+
+            int partition = currentPartition.get().number();
+            unspilledInputPages = spiller.map(spiller -> spiller.getSpilledPages(partition))
+                    .orElse(emptyIterator());
+
+            Optional.ofNullable(savedRows.remove(partition)).ifPresent(savedRow -> {
+                restoreProbe(
+                        savedRow.row,
+                        savedRow.joinPositionWithinPartition,
+                        savedRow.currentProbePositionProducedRow,
+                        savedRow.joinSourcePositions,
+                        SpillInfoSnapshot.noSpill());
+            });
+
+            return;
+        }
+
+        if (lookupPartitions.hasNext()) {
+            currentPartition.ifPresent(Partition::release);
+            currentPartition = Optional.of(lookupPartitions.next());
+            unspilledLookupSource = Optional.of(currentPartition.get().load());
+
+            return;
+        }
+
+        currentPartition.ifPresent(Partition::release);
+        if (lookupSourceProvider != null) {
```

Comment:
Maybe add a comment about this `if`.. `// There are no more lookup partitions to process, so clean up everything`